### PR TITLE
[TM-88] Location (Non-JavaScript)

### PR DESCRIPTION
**Pull request by mmiddleton3301**
Provides location select functionality for users who have disabled JavaScript.

### CODE REVIEW

File: src/SFA.DAS.LevyTransferMatching.Web/Validators/Location/LocationValidatorService.cs [hunk 1]
```diff
@@ -15,31 +16,52 @@ public LocationValidatorService(ILocationService locationService)
...
 
-            await CheckLocationsExist(errors, request.Locations);
+            await CheckLocationsExist(errors, request.Locations, multipleValidResults);
             CheckForDuplicates(errors, request.Locations);
 
             return errors;
         }
 
-        private async Task CheckLocationsExist(Dictionary<int, string> errors, List<string> locations)
+        private async Task CheckLocationsExist(Dictionary<int, string> errors, List<string> locations, IDictionary<int, IEnumerable<string>> multipleValidResults)
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Web/Validators/Location/LocationValidatorService.cs:29]: This method is too complex - could you see if it could be refactored? It's really difficult to follow. The intent should be clear enough that comments aren't required.

See also: SonarCloud assessment: `Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed.`

File: src/SFA.DAS.LevyTransferMatching.Web/Models/Cache/CreatePledgeCacheItem.cs [hunk 2]
```diff
@@ -25,5 +25,6 @@ public CreatePledgeCacheItem(Guid key)
         public List<string> JobRoles { get; set; }
         public List<string> Levels { get; set; }
         public List<string> Locations { get; set; }
+        public IDictionary<int, IEnumerable<string>> MultipleValidLocations { get; set; }
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Web/Models/Cache/CreatePledgeCacheItem.cs:28]: I don't think we should pollute this cache item with a property to hold this. Nothing against using the cache for this purpose (can't think of a viable alternative really), but I think a new cache item would be better. (See also, comment about a back link bug)

File: src/SFA.DAS.LevyTransferMatching.Web/Controllers/PledgesController.cs [hunk 3]
```diff
@@ -142,14 +143,45 @@ public async Task<IActionResult> Location(LocationRequest request)
...
+
             if (errors.Any())
             {
                 AddLocationErrorsToModelState(errors);
-                return RedirectToAction("Location", new { request.EncodedAccountId, request.AccountId, request.CacheKey });
+
+                return RedirectToAction(nameof(Location), new { request.EncodedAccountId, request.AccountId, request.CacheKey });
+            }
+
+            await _orchestrator.UpdateCacheItem(request);
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Web/Controllers/PledgesController.cs:157]: I think this leads to a bug whereby clicking the back link can result in invalid (ambiguous) locations being held in the cache item, which could then be saved to the pledge/db.

This cache item should only hold valid data at all times.

I _think_ this could be fixed by moving the update down below the next if{} block.

File: src/SFA.DAS.LevyTransferMatching.Web/Controllers/PledgesController.cs [hunk 4]
```diff
@@ -142,14 +143,45 @@ public async Task<IActionResult> Location(LocationRequest request)
...
+        {
+            var viewModel = await _orchestrator.GetLocationSelectViewModel(request);
+
+            return View(viewModel);
+        }
+
+        [Authorize]
+        [Route("create/location/select")]
+        [HttpPost]
+        public async Task<IActionResult> LocationSelect(LocationSelectPostRequest request)
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Web/Controllers/PledgesController.cs:181]: I haven't isolated it completely, but it looks like if you post this page without disambiguating the second of 2 ambigious locations, there is no link in the validation partial to take the user to the error.

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 5]
```diff
@@ -202,9 +202,38 @@ public async Task<ApplicationApprovedViewModel> GetApplicationApprovedViewModel(
             };
         }
         
-        public async Task<Dictionary<int, string>> ValidateLocations(LocationPostRequest request)
+        public async Task<LocationSelectViewModel> GetLocationSelectViewModel(LocationSelectRequest request)
         {
-            return await _validatorService.ValidateLocations(request);
+            var cacheItem = await RetrievePledgeCacheItem(request.CacheKey);
+
+            var locationSelectionGroups = cacheItem.MultipleValidLocations
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:209]: It's a bit hard to follow what this is doing

File: src/SFA.DAS.LevyTransferMatching.Web/Models/Pledges/LocationSelectPostRequest.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿namespace SFA.DAS.LevyTransferMatching.Web.Models.Pledges
+{
+    public class LocationSelectPostRequest : LocationSelectRequest
+    {
+        public LocationSelectionGroup[] LocationSelectionGroups { get; set; }
```
Comment (chrisfoster76) [inline src/SFA.DAS.LevyTransferMatching.Web/Models/Pledges/LocationSelectPostRequest.cs:5]: I think these properties could be named more clearly - perhaps to indicate that each "group" relates to an ambiguous location input, and each of the "items" are possible valid alternatives? I don't really get that from group/item.

**Review by chrisfoster76 — CHANGES_REQUESTED** [review]
A couple of bugs I noticed and code complexity issues.

File: src/SFA.DAS.LevyTransferMatching.Web/Validators/Location/LocationValidatorService.cs (hunk 1 above)
Comment (mmiddleton3301) [reply src/SFA.DAS.LevyTransferMatching.Web/Validators/Location/LocationValidatorService.cs]: Yeah, it was a bit of a mess to be fair. I've refactored this.

File: src/SFA.DAS.LevyTransferMatching.Web/Models/Cache/CreatePledgeCacheItem.cs (hunk 2 above)
Comment (mmiddleton3301) [reply src/SFA.DAS.LevyTransferMatching.Web/Models/Cache/CreatePledgeCacheItem.cs]: This has been split out into a `LocationSelectionCacheItem`, accessed via the same key with a prefix of the same name (e.g. `"LocationSelectionCacheItem_BE022A26-5568-4CC2-B459-226B7FE6E0C1"`), as discussed.

File: src/SFA.DAS.LevyTransferMatching.Web/Models/Pledges/LocationSelectPostRequest.cs (hunk 6 above)
Comment (mmiddleton3301) [reply src/SFA.DAS.LevyTransferMatching.Web/Models/Pledges/LocationSelectPostRequest.cs:5]: Renamed sub-classes to `SelectValidLocationGroup` and `ValidLocationItem` respectively. Not an easy one to name, tbh.

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs (hunk 5 above)
Comment (mmiddleton3301) [reply src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs]: I've refactored the LINQ statement a little in an attempt to improve code readability.

File: src/SFA.DAS.LevyTransferMatching.Web/Controllers/PledgesController.cs (hunk 4 above)
Comment (mmiddleton3301) [reply src/SFA.DAS.LevyTransferMatching.Web/Controllers/PledgesController.cs]: Thanks for your help narrowing down the fix for this. It's now been implemented, and has been unit tested.

**Review by chrisfoster76 — APPROVED** [review]
